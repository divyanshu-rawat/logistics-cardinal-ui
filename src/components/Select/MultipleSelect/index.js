import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import SelectPlus from 'react-select-plus';

import Select from '../../Select';

class MultipleSelect extends PureComponent {
  static propTypes = {
    name: PropTypes.string.isRequired,
    options: PropTypes.array.isRequired,
    validationState: PropTypes.oneOf([null, 'success', 'warning', 'error']),
    onChange: PropTypes.func.isRequired,
    placeholder: PropTypes.string,
    initialValue: PropTypes.array,
  };

  state = {
    selectedValue: this.props.initialValue,
  };

  componentDidMount() {
    const { initialValue } = this.props;

    if (initialValue) {
      this.onChange(initialValue);
    }
  }

  onChange = (option) => {
    this.setState({ selectedValue: option }, () =>
      this.props.onChange({ option }),
    );
  };

  render() {
    const { options, name, placeholder, validationState } = this.props;
    const { selectedValue } = this.state;

    return (
      <Select validationState={validationState}>
        <SelectPlus
          clearable={false}
          searchable={false}
          joinValues
          name={name}
          multi
          value={selectedValue}
          placeholder={placeholder}
          options={options}
          onChange={this.onChange}
        />
      </Select>
    );
  }
}

export default MultipleSelect;
