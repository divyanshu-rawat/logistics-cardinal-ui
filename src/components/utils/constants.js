export const SIZES = {
  lg: 'lg',
  md: 'md',
  sm: 'sm',
  xs: 'xs',
};
