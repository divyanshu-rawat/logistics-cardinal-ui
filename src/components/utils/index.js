export const calculatePercentage = (context, target) => target / context * 100;
