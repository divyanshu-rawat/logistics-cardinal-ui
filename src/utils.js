// import moment from 'moment-timezone';
// import cloneDeep from 'lodash/cloneDeep';
// import maxBy from 'lodash/maxBy';

// import { generateTimeCells } from 'containers/Scheduler/utils';
// import chartColors from './chartColors';
// import { capitalize } from '../../utils';

// export function prepareForecastDataStructure(dataObject, resultObject) {
//   const {
//     dataArray,
//     dateFieldName,
//     ordersFieldName,
//     dateCondition,
//     startTime,
//     endTime,
//   } = dataObject;
//   const {
//     charts,
//     tableValues,
//     weekDates,
//     timeIndexHelper,
//     colors,
//   } = resultObject;
//   let { biggestOrdersNumber } = resultObject;
//   dataArray.forEach((el) => {
//     const date = moment.utc(el[dateFieldName]).local();
//     const formattedDate = date.format('YYYY-MM-DD');
//     const formattedTime = date.format('HH:mm:ss');
//     const dayOfWeek = (date.day() || 7).toString();
//     // pattern WXX-YY
//     const weekOfYearString = `W${date.isoWeek()}-${date.format('YY')}`;
//     let tableDataLast = tableValues[tableValues.length - 1];

//     if (dateCondition(date)) {
//       return;
//     }

//     if (formattedTime < startTime || formattedTime > endTime) {
//       return;
//     }

//     if (!tableDataLast || tableDataLast.week !== weekOfYearString) {
//       tableValues.push({ week: weekOfYearString });
//       tableDataLast = tableValues[tableValues.length - 1];
//     }

//     if (!tableDataLast[dayOfWeek]) {
//       tableDataLast[dayOfWeek] = 0;
//     }

//     tableDataLast[dayOfWeek] += el[ordersFieldName];
//     if (tableDataLast[dayOfWeek] > biggestOrdersNumber) {
//       biggestOrdersNumber = tableDataLast[dayOfWeek];
//     }

//     if (!charts[dayOfWeek]) {
//       charts[dayOfWeek] = {
//         data: [{ formattedTime }],
//         dates: {},
//       };
//       timeIndexHelper[dayOfWeek] = {
//         [formattedTime]: 0,
//       };
//     }

//     if (!colors[weekOfYearString]) {
//       colors[weekOfYearString] = chartColors[tableValues.length - 1];
//     }

//     if (timeIndexHelper[dayOfWeek][formattedTime] === undefined) {
//       charts[dayOfWeek].data.push({ formattedTime });
//       timeIndexHelper[dayOfWeek][formattedTime] =
//         charts[dayOfWeek].data.length - 1;
//     }

//     const index = timeIndexHelper[dayOfWeek][formattedTime];
//     charts[dayOfWeek].data[index][formattedDate] = el[ordersFieldName];
//     charts[dayOfWeek].dates[formattedDate] = {
//       formattedDate,
//       lineColor: chartColors[tableValues.length - 1],
//     };

//     if (weekDates[weekOfYearString] === undefined) {
//       weekDates[weekOfYearString] = [];
//     }

//     if (weekDates[weekOfYearString].indexOf(formattedDate) === -1) {
//       weekDates[weekOfYearString].push(formattedDate);
//     }
//   });

//   Object.keys(charts).forEach((day) => {
//     charts[day].data = charts[day].data.sort(
//       (a, b) => (a.formattedTime < b.formattedTime ? -1 : 1),
//     );
//   });

//   return {
//     charts,
//     tableValues,
//     weekDates,
//     timeIndexHelper,
//     biggestOrdersNumber,
//     colors,
//   };
// }

// export function calculateChange(value, prevValue, index) {
//   // change way of storing data, create an object with needed information
//   const result = {
//     value,
//   };
//   // if there are no data provided at all it is hard to calculate
//   if (value !== undefined && prevValue !== undefined) {
//     // for first week we have no data to compare
//     if (index) {
//       result.change = Math.round((value - prevValue) / prevValue * 100);
//       // in the case if prevValue is 0, we assume change is 100% or 0%, depends on current value
//       if (!prevValue) {
//         result.change = value ? 100 : 0;
//       }
//     }
//   }

//   return result;
// }

// /**
//  * Function for calculation of percent change in orders comparing to the same day in previous week
//  * Also sums orders per week for total column
//  * @param data Array of data, where every element is a week
//  * @param days Array od days numbers
//  * @returns Array, the clone of data with additional information for every day
//  */
// export function calculateOrders(data, days) {
//   // Never mutate original data
//   const clonedData = cloneDeep(data);
//   // iterate through every week
//   clonedData.forEach((weekData, index) => {
//     // object with whole week data, in which changes are made
//     const weekObj = weekData;
//     // iterate through days
//     days.forEach((day, i) => {
//       // order value is stored in data[index][day]
//       const orders = weekData[day];
//       weekObj[day] = calculateChange(
//         orders,
//         index ? clonedData[index - 1][day].value : undefined,
//         index,
//       );

//       if (!weekObj.total) {
//         weekObj.total = 0;
//       }
//       weekObj.total += orders || 0;

//       // if all days in the week were calculated
//       if (i === days.length - 1) {
//         weekObj.total = calculateChange(
//           weekObj.total,
//           index ? clonedData[index - 1].total.value : undefined,
//           index,
//         );
//       }
//     });
//     clonedData[index] = weekObj;
//   });
//   return clonedData;
// }

// export function createForecastDataStructure(dataObject, filters) {
//   let results = {
//     charts: {},
//     tableValues: [],
//     weekDates: {},
//     timeIndexHelper: {},
//     biggestOrdersNumber: 0,
//     biggestTotalValue: 0,
//     colors: {},
//   };

//   if (!dataObject.forecasts || !dataObject.actuals) {
//     delete results.timeIndexHelper;
//     return results;
//   }
//   const forecasts = dataObject.forecasts.sort(
//     (a, b) => (a.forecast_for < b.forecast_for ? -1 : 1),
//   );
//   const actuals = dataObject.actuals.sort(
//     (a, b) => (a.ordered_at < b.ordered_at ? -1 : 1),
//   );
//   const borderDate = actuals[actuals.length - 1].ordered_at;
//   const { startTime, endTime } = filters;

//   {
//     const actualsDateCondition = (date) =>
//       date.isAfter(moment.parseZone(borderDate), 'day');
//     results = prepareForecastDataStructure(
//       {
//         dataArray: actuals,
//         dateFieldName: 'ordered_at',
//         ordersFieldName: 'orders',
//         dateCondition: actualsDateCondition,
//         startTime,
//         endTime,
//       },
//       results,
//     );
//   }

//   {
//     const forecastsDateCondition = (date) =>
//       date.isSameOrBefore(moment.parseZone(borderDate), 'day');
//     results = prepareForecastDataStructure(
//       {
//         dataArray: forecasts,
//         dateFieldName: 'forecast_for',
//         ordersFieldName: 'adjusted_orders',
//         dateCondition: forecastsDateCondition,
//         startTime,
//         endTime,
//       },
//       results,
//     );
//   }

//   results.tableValues = calculateOrders(
//     results.tableValues,
//     Object.keys(results.charts),
//   ).reverse();
//   results.biggestTotalValue = maxBy(
//     results.tableValues,
//     (week) => week.total.value,
//   ).total.value;

//   delete results.timeIndexHelper;
//   return results;
// }

// export function formatZoneList(zones) {
//   return zones.map((zone) => {
//     const value = zone.id.replace('zone:', '');
//     return {
//       value,
//       label: capitalize(zone.fields.name),
//     };
//   });
// }

// export function updateSelectedDays(tableData, selectedDays) {
//   const updatedSelectedDays = cloneDeep(selectedDays);
//   tableData.forEach((weekData) => {
//     selectedDays.forEach((dayData, index) => {
//       if (weekData.week === dayData.dataObj.week) {
//         updatedSelectedDays[index].dataObj.value =
//           weekData[dayData.dataObj.day].value;
//       }
//     });
//   });

//   return updatedSelectedDays;
// }

// export function createTimeOptions(optionsObj = {}) {
//   const labels = generateTimeCells({
//     startHour: 0,
//     endHour: 23,
//     startSeconds: 0,
//     startMinutes: 0,
//     endMinutes: 59,
//     endSeconds: 59,
//     interval: 30,
//     format: 'HH:mm',
//     ...optionsObj,
//   });
//   const options = labels.map((label) => ({ label, value: `${label}:00` }));
//   options.push({ label: '23:59', value: '23:59:59' });
//   return options;
// }

// export function createAdjustmentDataObject(dataObj) {
//   const { adjustment, filters } = dataObj;
//   return {
//     zone_ids: filters.zoneIds,
//     forecast_dates: adjustment.selectedDays.map((date) =>
//       date.date.format('YYYY-MM-DD'),
//     ),
//     start_time: filters.startTime,
//     end_time: filters.endTime,
//     adjust_by: adjustment.adjustment,
//     applies_to: adjustment.appliesTo.map(
//       (option) => (option.value ? option.value : option),
//     ),
//     fixed: adjustment.fixAdjustment,
//     reason: adjustment.reason,
//     comment: adjustment.reasonDetail,
//   };
// }

// export function updateHistory(history, adjustment) {
//   const newHistory = cloneDeep(history);
//   history.some((row, index) => {
//     if (row.id === adjustment.id) {
//       newHistory[index] = adjustment;
//       return true;
//     }
//     return false;
//   });
//   return newHistory;
// }
