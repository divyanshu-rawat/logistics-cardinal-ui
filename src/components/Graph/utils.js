import { SIZES } from '../utils/constants';

/**
 * Generate random shifts data
 * @returns {{shifts: number}[]}
 */
export function generateRandomShifts() {
  const MONTHS = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec',
  ];
  return [...new Array(12)].map((_, index) => ({
    shifts: Math.round(Math.random() * 50),
    month: MONTHS[index],
  }));
}

export function getStrokeWidth(size) {
  let strokeSize;

  switch (size) {
    case SIZES.lg:
      strokeSize = 4;
      break;
    case SIZES.md:
      strokeSize = 3;
      break;
    case SIZES.sm:
      strokeSize = 2;
      break;
    default:
      strokeSize = 1;
  }

  return strokeSize;
}
