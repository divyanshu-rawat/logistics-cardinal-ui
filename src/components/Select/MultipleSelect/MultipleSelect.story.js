import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';
import centered from '@storybook/addon-centered';

import withTests from '../../../utils/withTests';

import { MultipleSelectMock } from '../../../utils/mocks';
import MultipleSelect from '.';

storiesOf('Select/MultipleSelect', module)
  .addDecorator(withTests('index'))
  .addDecorator(centered)
  .add(
    'Normal',
    withInfo()(() => (
      <MultipleSelect
        name="zones"
        onChange={action((item) => console.log(item))}
        options={MultipleSelectMock}
      />
    )),
  )
  .add(
    'With initial Values',
    withInfo()(() => (
      <MultipleSelect
        name="zones"
        initialValue={[{ value: 'stockholm$stockholm', label: 'Stockholm' }]}
        onChange={action((item) => console.log(item))}
        options={MultipleSelectMock}
      />
    )),
  );
