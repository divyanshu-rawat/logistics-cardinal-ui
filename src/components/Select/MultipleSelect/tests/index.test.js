import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { MultipleSelectMock } from '../../../../utils/mocks';
import MultipleSelect from '..';

const onChangeFn = jest.fn();

const renderedComponent = (value) =>
  shallow(
    <MultipleSelect
      options={MultipleSelectMock}
      name="zones"
      placeholder="Select Zone..."
      onChange={onChangeFn}
      {...(value ? { initialValue: value } : {})}
    />,
  );

describe('<MultipleSelect />', () => {
  it('should render the component', () => {
    const tree = toJson(renderedComponent());
    expect(tree).toMatchSnapshot();
    expect(onChangeFn.mock.calls.length).toBe(0);
  });

  it('should init component with given `initialValue` prop', () => {
    onChangeFn.mockClear();
    const component = renderedComponent([MultipleSelectMock[0]]);
    const tree = toJson(component);

    expect(onChangeFn.mock.calls.length).toBe(1);
    expect(tree).toMatchSnapshot();

    onChangeFn.mockClear();
    component.setProps({
      initialValue: [MultipleSelectMock[1]],
    });

    expect(onChangeFn.mock.calls.length).toBe(1);
  });

  it('should not call onChange if initialValue is still equal', () => {
    const component = renderedComponent([MultipleSelectMock[0]]);
    onChangeFn.mockClear();

    component.setProps({
      initialValue: [MultipleSelectMock[0]],
    });

    expect(onChangeFn.mock.calls.length).toBe(0);
  });

  it('MultipleSelect: should call `onChangeFn` when value is changed', () => {
    onChangeFn.mockClear();
    renderedComponent()
      .find('Select')
      .simulate('change', { target: { value: [MultipleSelectMock[1]] } });

    expect(onChangeFn.mock.calls.length).toBe(1);
  });
});
