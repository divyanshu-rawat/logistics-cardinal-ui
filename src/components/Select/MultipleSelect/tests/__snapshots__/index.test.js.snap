// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultipleSelect /> should init component with given \`initialValue\` prop 1`] = `
<styled.div>
  <Select
    arrowRenderer={[Function]}
    autosize={true}
    backspaceRemoves={true}
    backspaceToRemoveMessage="Press backspace to remove {label}"
    clearAllText="Clear all"
    clearRenderer={[Function]}
    clearValueText="Clear value"
    clearable={false}
    closeOnSelect={true}
    deleteRemoves={true}
    delimiter=","
    disabled={false}
    dropdownComponent={[Function]}
    escapeClearsValue={true}
    filterOptions={[Function]}
    ignoreAccents={true}
    ignoreCase={true}
    inputProps={Object {}}
    isLoading={false}
    joinValues={true}
    labelKey="label"
    matchPos="any"
    matchProp="any"
    menuBuffer={0}
    menuRenderer={[Function]}
    multi={true}
    name="zones"
    noResultsText="No results found"
    onBlurResetsInput={true}
    onChange={[Function]}
    onCloseResetsInput={true}
    onSelectResetsInput={true}
    openOnClick={true}
    optionComponent={[Function]}
    optionGroupComponent={[Function]}
    options={
      Array [
        Object {
          "label": "Stockholm",
          "value": "stockholm$stockholm",
        },
        Object {
          "label": "Solna",
          "value": "stockholm$solna",
        },
        Object {
          "label": "Validation_zone",
          "value": "stockholm$validation_zone",
        },
      ]
    }
    pageSize={5}
    placeholder="Select Zone..."
    removeSelected={true}
    required={false}
    rtl={false}
    scrollMenuIntoView={true}
    searchable={false}
    simpleValue={false}
    tabSelectsValue={true}
    trimFilter={true}
    value={
      Array [
        Object {
          "label": "Stockholm",
          "value": "stockholm$stockholm",
        },
      ]
    }
    valueComponent={[Function]}
    valueKey="value"
  />
</styled.div>
`;

exports[`<MultipleSelect /> should render the component 1`] = `
<styled.div>
  <Select
    arrowRenderer={[Function]}
    autosize={true}
    backspaceRemoves={true}
    backspaceToRemoveMessage="Press backspace to remove {label}"
    clearAllText="Clear all"
    clearRenderer={[Function]}
    clearValueText="Clear value"
    clearable={false}
    closeOnSelect={true}
    deleteRemoves={true}
    delimiter=","
    disabled={false}
    dropdownComponent={[Function]}
    escapeClearsValue={true}
    filterOptions={[Function]}
    ignoreAccents={true}
    ignoreCase={true}
    inputProps={Object {}}
    isLoading={false}
    joinValues={true}
    labelKey="label"
    matchPos="any"
    matchProp="any"
    menuBuffer={0}
    menuRenderer={[Function]}
    multi={true}
    name="zones"
    noResultsText="No results found"
    onBlurResetsInput={true}
    onChange={[Function]}
    onCloseResetsInput={true}
    onSelectResetsInput={true}
    openOnClick={true}
    optionComponent={[Function]}
    optionGroupComponent={[Function]}
    options={
      Array [
        Object {
          "label": "Stockholm",
          "value": "stockholm$stockholm",
        },
        Object {
          "label": "Solna",
          "value": "stockholm$solna",
        },
        Object {
          "label": "Validation_zone",
          "value": "stockholm$validation_zone",
        },
      ]
    }
    pageSize={5}
    placeholder="Select Zone..."
    removeSelected={true}
    required={false}
    rtl={false}
    scrollMenuIntoView={true}
    searchable={false}
    simpleValue={false}
    tabSelectsValue={true}
    trimFilter={true}
    value={Array []}
    valueComponent={[Function]}
    valueKey="value"
  />
</styled.div>
`;
